README.md
<!--
SPDX-FileCopyrightText: 2020 Dimitris Lampridis <dlampridis@gmail.com>

SPDX-License-Identifier: CC0-1.0
-->

# DeviMote
Unofficial remote control for Devialet Expert amplifiers written in Python

![Pylint](https://github.com/gnulabis/devimote/workflows/Pylint/badge.svg?branch=main)
![REUSE](https://github.com/gnulabis/devimote/workflows/REUSE/badge.svg?branch=main)

![](doc/images/devimote_demo.gif)


CHANGELOG.rst
..
   SPDX-FileCopyrightText: 2020 Dimitris Lampridis <dlampridis@gmail.com>
   SPDX-License-Identifier: CC0-1.0

=========
Changelog
=========

Unreleased
==========

Added
-----
- First version


wireshark/dec.lua
-- SPDX-FileCopyrightText: 2020 Dimitris Lampridis <dlampridis@gmail.com>
--
-- SPDX-License-Identifier: GPL-3.0-or-later

dec_protocol = Proto("DEC",  "Devialet Expert Control")

dec_protocol.fields.magic_val   = ProtoField.string("dec.magic",      "Magic Value")
dec_protocol.fields.packet_cnt  = ProtoField.uint16("dec.packet_cnt", "Packet Counter")
dec_protocol.fields.cmd_cnt     = ProtoField.uint16("dec.cmd_cnt",    "Command Counter")
dec_protocol.fields.cmd_type    = ProtoField.string("dec.cmd_type",   "Command")
dec_protocol.fields.out_sel     = ProtoField.uint16("dec.out_sel",    "Output Select")
dec_protocol.fields.volume      = ProtoField.uint16("dec.volume",     "Volume", base.HEX)
dec_protocol.fields.crc16_check = ProtoField.string("dec.crc_check",  "Checksum")

dec_protocol.fields.todo_field = ProtoField.none("dec.todo", "Unknown Field")
dec_protocol.fields.zero_field = ProtoField.none("dec.zero", "Zero Padding")

function dec_protocol.dissector(buffer, pinfo, tree)

   length = buffer:len()
   if length == 0 then return end

   pinfo.cols.protocol = dec_protocol.name

   local subtree = tree:add(dec_protocol, buffer(), dec_protocol.description)
   subtree:add(dec_protocol.fields.magic_val,   buffer(0x0,2))
   subtree:add(dec_protocol.fields.packet_cnt,  buffer(0x2,2))
   subtree:add(dec_protocol.fields.cmd_cnt,     buffer(0x4,2))

   local opstring = "UNKNOWN"
   local opcode = buffer(0x6,2):uint()

   if     (opcode == 0x0004) then opstring = "VOLUME"
   elseif (opcode == 0x0005) then opstring = "OUTPUT"
   elseif (opcode == 0x0001) then opstring = "STANDBY"
   elseif (opcode == 0x0101) then opstring = "ON"
   elseif (opcode == 0x0107) then opstring = "MUTE"
   elseif (opcode == 0x0007) then opstring = "UNMUTE"
   end

   subtree:add(dec_protocol.fields.cmd_type, buffer(0x6,2), opstring)

   if (opstring == "VOLUME") then
      local volume = buffer(0x8,2):uint()
      subtree:add(dec_protocol.fields.volume, volume)
      subtree:add(dec_protocol.fields.zero_field, buffer(0xa,2))
      pinfo.cols.info:append(string.format(" Cmd=%s %x", opstring, volume))
   elseif (opstring == "OUTPUT SELECT") then
      local out_sel = bit32.band(buffer(0x8,2):uint(), 0x01ff)
      if (bit32.band(out_sel, 0x100) ~= 0) then
         out_sel = bit32.band(out_sel, 0xff)
         out_sel = bit32.lshift(out_sel, 1)
         out_sel = bit32.bor(out_sel, 0x100)
      end
      out_sel = bit32.rshift(out_sel, 5)
      subtree:add(dec_protocol.fields.out_sel, buffer(0x8,2), out_sel)
      subtree:add(dec_protocol.fields.zero_field, buffer(0xa,2))
      pinfo.cols.info:append(string.format(" Cmd=%s #%02d", opstring, out_sel))
   else
      subtree:add(dec_protocol.fields.zero_field, buffer(0x8,4))
      pinfo.cols.info:append(string.format(" Cmd=%s", opstring))
   end

   -- ccitt_16() taken from https://github.com/clarkli86/crc16_ccitt
   -- loop adjusted for wireshark Tvbrange as argument
   local function ccitt_16(byte_array)

      local POLY = 0x1021

      local function hash(crc, byte)
         for i = 0, 7 do
            local bit = bit32.extract(byte, 7 - i) -- Take the lsb
            local msb = bit32.extract(crc, 15, 1) -- msb
            crc = bit32.lshift(crc, 1) -- Remove the lsb of crc
            if bit32.bxor(bit, msb) == 1 then crc = bit32.bxor(crc, POLY) end
         end
         return crc
      end

      local crc = 0xffff
      for i = 0, byte_array:len()-1, 1 do
         crc = hash(crc, byte_array:range(i,1):uint())
      end

      return bit32.extract(crc, 0, 16)
   end

   local crc16_calc = ccitt_16(buffer(0x0,12))
   local crc16_read = buffer(0xc,2):uint()
   subtree:add(dec_protocol.fields.crc16_check, buffer(0xc,2), crc16_calc == crc16_read and "Valid" or "Invalid")

   subtree:add(dec_protocol.fields.todo_field, buffer(0xe,128))

end

local udp_port = DissectorTable.get("udp.port")
udp_port:add(45455, dec_protocol)


wireshark/des.lua
-- SPDX-FileCopyrightText: 2020 Dimitris Lampridis <dlampridis@gmail.com>
--
-- SPDX-License-Identifier: GPL-3.0-or-later

des_protocol = Proto("DES",  "Devialet Expert Status")

des_protocol.fields.magic_val   = ProtoField.string("des.magic",      "Magic Value")
des_protocol.fields.packet_cnt  = ProtoField.uint16("des.packet_cnt", "Packet Counter")
des_protocol.fields.dev_name    = ProtoField.string("des.dev_name",   "Device Name")
des_protocol.fields.dev_uptime  = ProtoField.string("des.dev_uptime", "Uptime")
des_protocol.fields.crc16_check = ProtoField.string("des.crc_check",  "Checksum")
des_protocol.fields.status      = ProtoField.string("des.status",     "Status")
des_protocol.fields.volume      = ProtoField.string("des.volume",     "Volume")

des_protocol.fields.todo_field1 = ProtoField.none("des.todo1", "Unknown Field")
des_protocol.fields.todo_field2 = ProtoField.none("des.todo2", "Unknown Field")
des_protocol.fields.todo_field3 = ProtoField.none("des.todo3", "Unknown Field")
des_protocol.fields.todo_field4 = ProtoField.none("des.todo4", "Unknown Field")
des_protocol.fields.todo_field5 = ProtoField.none("des.todo5", "Unknown Field")

output_channel = {}

for i = 0, 14, 1 do
   output_channel[i] = ProtoField.string(string.format("des.out%d", i), string.format("Output #%02d", i))
   table.insert(des_protocol.fields, output_channel[i])
end


function des_protocol.dissector(buffer, pinfo, tree)

   length = buffer:len()
   if length == 0 then return end

   pinfo.cols.protocol = des_protocol.name

   local subtree = tree:add(des_protocol, buffer(), des_protocol.description)
   subtree:add(des_protocol.fields.magic_val,   buffer(0x0,2))
   subtree:add(des_protocol.fields.packet_cnt,  buffer(0x2,2))
   subtree:add(des_protocol.fields.todo_field1, buffer(0x4,15))
   subtree:add(des_protocol.fields.dev_name,    buffer(0x013,31))
   subtree:add(des_protocol.fields.todo_field2, buffer(0x32,2))

   for i = 0, 14, 1 do
      local out_val = string.format("%s [%s]", buffer(0x35+17*i, 16):string(), buffer(0x34+17*i, 1):uint() ~= 0x30 and "Enabled" or "Disabled")
      subtree:add(output_channel[i], buffer(0x34+17*i,17), out_val)
   end

   local function disp_status(stat)
      local power    = (bit32.band(stat, 0x8000) ~= 0) and "On" or "Standby"
      local muted    = (bit32.band(stat, 0x0002) ~= 0) and "Yes" or "No"
      local out_num  = bit32.rshift(bit32.band(stat, 0x003c), 2)
      local out_name = buffer(0x35+17*out_num, 16):string()
      return string.format("Power: %s, Output: %s, Muted: %s", power, out_name, muted)
   end

   subtree:add(des_protocol.fields.status, buffer(0x133,2), disp_status(buffer(0x133,2):uint()))

   subtree:add(des_protocol.fields.todo_field3, buffer(0x135,1))

   local vol = (buffer(0x136,1):uint() - 195) / 2.0
   subtree:add(des_protocol.fields.volume, buffer(0x136,1), string.format("%02.1f dB", vol))

   subtree:add(des_protocol.fields.todo_field4, buffer(0x137,20))

   local function disp_uptime(time)
      local days = math.floor(time / 86400000)
      local hours = math.floor((time % 86400000)/3600000)
      local minutes = math.floor((time % 3600000)/60000)
      local seconds = math.floor((time % 60000)/1000)
      local milli = math.floor(time % 1000)
      return string.format("%d days, %02d hours, %02d minutes, %02d sec, %03d ms", days, hours, minutes, seconds, milli)
   end

   subtree:add(des_protocol.fields.dev_uptime, buffer(0x014b,4), disp_uptime(buffer(0x014b,4):uint()))

   -- ccitt_16() taken from https://github.com/clarkli86/crc16_ccitt
   -- loop adjusted for wireshark Tvbrange as argument
   local function ccitt_16(byte_array)

      local POLY = 0x1021

      local function hash(crc, byte)
         for i = 0, 7 do
            local bit = bit32.extract(byte, 7 - i) -- Take the lsb
            local msb = bit32.extract(crc, 15, 1) -- msb
            crc = bit32.lshift(crc, 1) -- Remove the lsb of crc
            if bit32.bxor(bit, msb) == 1 then crc = bit32.bxor(crc, POLY) end
         end
         return crc
      end

      local crc = 0xffff
      for i = 0, byte_array:len()-1, 1 do
         crc = hash(crc, byte_array:range(i,1):uint())
      end

      return bit32.extract(crc, 0, 16)
   end

   subtree:add(des_protocol.fields.todo_field5, buffer(0x14f,8))

   local crc16_calc = ccitt_16(buffer(0x0,343))
   local crc16_read = buffer(0x157,2):uint()
   subtree:add(des_protocol.fields.crc16_check, buffer(0x157,2), crc16_calc == crc16_read and "Valid" or "Invalid")
end

local udp_port = DissectorTable.get("udp.port")
udp_port:add(45454, des_protocol)


src/devimote.py
# SPDX-FileCopyrightText: 2020 Dimitris Lampridis <dlampridis@gmail.com>
#
# SPDX-License-Identifier: GPL-3.0-or-later

'''An unofficial remote control application for Devialet Expert amplifiers'''

import socket
import struct
import math as m

from kivy.app import App
from kivy.properties import ObjectProperty # pylint: disable=no-name-in-module
from kivy.clock import Clock
from kivy.uix.gridlayout import GridLayout
from kivy.uix.boxlayout import BoxLayout

class DeviMoteVolume(BoxLayout):
    '''Wrapper class around BoxLayout'''
    vol_slider: ObjectProperty(None)

    def set_byte (self, byte):
        '''Function to adjust the volume slider'''
        self.vol_slider.value = byte

class DeviMoteWidget(GridLayout):
    '''Top-level widget'''
    channels  = ObjectProperty(None)
    volume    = ObjectProperty(None)
    stat_line = ObjectProperty(None)
    sw_power  = ObjectProperty(None)
    sw_mute   = ObjectProperty(None)

    def set_volume(self, byte):
        '''Function to adjust the volume'''
        self.volume.set_byte(byte)

    def update(self, status):
        '''Function to update all GUI elements based on current status'''
        if status and status['connected']:
            self.stat_line.text = 'Status: Connected'
            if status['power']:
                self.sw_power.state = 'down'
                self.sw_power.text  = 'ON'
                self.sw_power.background_color = [1, 1, 1, 1]
            elif status['booting']:
                self.sw_power.state = 'down'
                self.sw_power.text  = 'BOOTING'
                self.sw_power.background_color = [.25, .25, .25, 1]
            else:
                self.sw_power.state = 'normal'
                self.sw_power.text  = 'STANDBY'
                self.sw_power.background_color = [1, 1, 1, 1]
            self.sw_mute.state  = 'down' if status['muted'] else 'normal'
            self.volume.set_byte(status['volume'])
            if not self.channels.values:
                for channel in status['ch_list']:
                    self.channels.values.append(status['ch_list'][channel])
            self.channels.text = status['ch_list'][status['channel']]
        else:
            self.stat_line.text  = 'Status: Not connected'

def _crc16(data : bytearray):
    '''Internal function to calculate a CRC-16/CCITT-FALSE from the given bytearray'''
    if data is None :
        return 0
    crc = 0xFFFF
    for i in enumerate(data):
        crc ^= data[i[0]] << 8
        for _ in range(8):
            if (crc & 0x8000) > 0:
                crc =(crc << 1) ^ 0x1021
            else:
                crc = crc << 1
    return crc & 0xFFFF

class DeviMoteBackEnd():
    '''Backend class handling all control and status monitoring'''
    UDP_PORT_STATUS = 45454
    UDP_PORT_CMD    = 45455
    VOLUME_LIMIT    = -10

    def __init__(self):
        '''Backend constructor with default initial values'''
        self.status = {}
        self.status['dev_name'] = 'Unknown'
        self.status['ip'] = None
        self.status['ch_list']  = {}
        self.status['power'] = False
        self.status['muted'] = False
        self.status['channel'] = 0
        self.status['volume'] = 0
        self.status['connected'] = False
        self.status['crc_ok'] = False
        self.packet_cnt = 0

    def _send_command(self, data: bytearray):
        '''Internal function that builds and transmits a UDP packet command to the amplifier'''
        if not (self.status['connected'] and self.status['ip']):
            return
        sock = socket.socket(socket.AF_INET,    # Internet
                             socket.SOCK_DGRAM) # UDP
        data[0] = 0x44
        data[1] = 0x72
        for _ in range(4):
            data[3] = self.packet_cnt
            data[5] = self.packet_cnt >> 1
            self.packet_cnt += 1
            crc = _crc16(data[0:12])
            data[12] = (crc & 0xff00) >> 8
            data[13] = (crc & 0x00ff)
            sock.sendto(data, (self.status['ip'], self.UDP_PORT_CMD))

    def toggle_power(self):
        '''Function for toggling the power status'''
        data = bytearray(142)
        data[6] = int(not self.status['power'])
        data[7] = 0x01
        self._send_command(data)

    def toggle_mute(self):
        '''Function for toggling the mute status'''
        data = bytearray(142)
        data[6] = int(not self.status['muted'])
        data[7] = 0x07
        self._send_command(data)

    def set_volume(self, db_value):
        '''Function for changing the volume'''

        if db_value > self.VOLUME_LIMIT:
            db_value = self.VOLUME_LIMIT

        def _db_convert(db_value):
            '''Internal function to convert dB to a 16-bit representation used by set_volume'''
            db_abs = m.fabs(db_value)
            if db_abs == 0:
                retval = 0
            elif db_abs == 0.5:
                retval = 0x3f00
            else:
                retval = (256 >> m.ceil(1 + m.log(db_abs, 2))) + _db_convert(db_abs - 0.5)
            return retval

        volume = _db_convert(db_value)

        if db_value < 0:
            volume |= 0x8000

        data = bytearray(142)
        data[6] = 0x00
        data[7] = 0x04
        data[8] = (volume & 0xff00) >> 8
        data[9] = (volume & 0x00ff)
        self._send_command(data)

    def set_output(self, output):
        '''Function for changing the output'''
        out_val = 0x4000 | (output << 5)
        data = bytearray(142)
        data[6] = 0x00
        data[7] = 0x05
        data[8] = (out_val & 0xff00) >> 8
        if output > 7:
            data[9] = (out_val & 0x00ff) >> 1
        else:
            data[9] = (out_val & 0x00ff)
        self._send_command(data)

    def update(self):
        '''Try to get UDP status packet and decode it'''
        sock = socket.socket(socket.AF_INET,    # Internet
                             socket.SOCK_DGRAM) # UDP
        sock.bind(('', self.UDP_PORT_STATUS))
        sock.settimeout(2)
        try:
            data, addr = sock.recvfrom(512) # buffer size is 512 bytes
        except socket.timeout:
            self.status['connected'] = False
            return self.status
        self.status['connected'] = True
        self.status['ip'] = addr[0]
        self.status['dev_name'] = data[19:50].decode('UTF-8')
        for i in range(0,15):
            enabled = int(chr(data[52+i*17]))
            if enabled:
                self.status['ch_list'][i] = data[53+i*17:52+(i+1)*17].decode('UTF-8')
        self.status['power']   = (data[307] & 0x80) != 0
        self.status['muted']   = (data[308] & 0x2) != 0
        self.status['channel'] = (data[308] & 0x3c) >> 2
        self.status['volume']  =  data[310]
        self.status['crc_ok']  = (_crc16(data[:-2]) == struct.unpack('>H',data[-2:])[0])

        return self.status

class DeviMoteApp(App):
    '''Top-level class combining the backend and the top-level widget'''
    def __init__(self):
        '''Constructor'''
        super().__init__()
        self.gui = None
        self.backend = None
        self.status = None

    def _powered(self, _dt):
        '''Internal function to use during booting'''
        self.status['booting'] = False

    def toggle_power_callback(self, _instance):
        '''Callback function for toggling power'''
        if self.status['booting']:
            return
        if not self.status['power']:
            self.status['booting'] = True
            Clock.schedule_once(self._powered, 20)
        self.backend.toggle_power()

    def toggle_mute_callback(self, _instance):
        '''Callback function for toggling mute'''
        self.backend.toggle_mute()

    def set_volume_callback(self, _instance, value):
        '''Callback function for changing the volume'''
        if value != self.status['volume']:
            self.backend.set_volume((value-195.0) / 2)

    def set_output_callback(self, _instance, text):
        '''Callback function for changing the output'''
        for channel in self.status['ch_list']:
            if text == self.status['ch_list'][channel]:
                output = channel
                break
        if output != self.status['channel']:
            self.backend.set_output(output)

    def build(self):
        '''Kivy build function, runs once'''
        self.gui = DeviMoteWidget()
        self.backend = DeviMoteBackEnd()
        self.status = self.backend.update()
        self._powered(0)
        self.gui.update(self.status)
        self.gui.sw_power.bind(on_press=self.toggle_power_callback)
        self.gui.sw_mute.bind(on_press=self.toggle_mute_callback)
        self.gui.volume.vol_slider.bind(value=self.set_volume_callback)
        self.gui.channels.bind(text=self.set_output_callback)
        Clock.schedule_interval(self.update, 0.1)
        return self.gui

    def update(self, _dt):
        '''Function to update both the backend and the GUI. Scheduled to run periodically'''
        self.status = self.backend.update()
        if self.status['power']:
            self._powered(0)
        self.gui.update(self.status)

    def report(self):
        '''Pretty-print current status'''
        if not self.status['crc_ok']:
            print ('[CRC ERROR]')
            return
        if self.status['connected']:
            print (
                "[{}] {} ({}), volume: {}dB {} {}".format(
                    "ON " if self.status['power'] else "OFF",
                    self.status['dev_name'],
                    self.status['ip'],
                    (self.status['volume'] - 195) / 2.0,
                    self.status['ch_list'][self.status['channel']],
                    "[M]" if self.status['muted'] else ""
                )
            )

if __name__ == '__main__':
    DeviMoteApp().run()
